pyquil/api/_compiler_client.py

```diff
@@ -0,0 +1,242 @@
+##############################################################################
+# Copyright 2016-2021 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Iterator, Optional, cast, List
+
+import rpcq
+from qcs_api_client.client import QCSClientConfiguration
+from rpcq.messages import TargetDevice as TargetQuantumProcessor
```

Comment:
Really nice to have all of the request and response shapes typed out here. I think this all belongs with the definitions in the pyquil.external.rpcq and we could eventually move all of this out into a separate compiler client library that Pyquil could depend on. If we end up doing that, my feeling is that it should have it's own configuration class that the QCS configuration could produce (without necessarily passing the entire QCS configuration). https://github.com/rigetti/pyquil/issues/1307

Reply:
The goal of this client is to actually be RPC-agnostic, from the outside. One day, we could refactor it internally to communicate over http. So, I wouldn't want to move this client and these definitions to an RPC-related module.

I do agree that there could be a future where the compiler client is a separate library, sort of like the QCS client